File: app/src/main/java/torille/fi/lurkforreddit/data/models/jsonResponses/Image.java
package torille.fi.lurkforreddit.data.models.jsonResponses;


import com.google.auto.value.AutoValue;
import com.google.gson.Gson;
import com.google.gson.TypeAdapter;
import com.google.gson.annotations.SerializedName;

import java.util.List;

/**
 * Model containing a single images resolutions and source image
 */
@AutoValue
public abstract class Image {

    @SerializedName("source")
    public abstract ImageSource source();

    @SerializedName("resolutions")
    public abstract List<ImageResolution> resolutions();

    @SerializedName("id")
    public abstract String id();

    public static TypeAdapter<Image> typeAdapter(Gson gson) {
        return new AutoValue_Image.GsonTypeAdapter(gson);

    }
}

File: app/src/main/java/torille/fi/lurkforreddit/data/models/jsonResponses/Image.kt
package torille.fi.lurkforreddit.data.models.jsonResponses


import com.google.gson.annotations.SerializedName

/**
 * Model containing a single images resolutions and source image
 */
data class Image(
    @SerializedName("source")
    val source: ImageSource,
    @SerializedName("resolutions")
    val resolutions: List<ImageResolution>,
    @SerializedName("subId")
    val id: String
)

/**
 * Model containg the source image
 */
data class ImageSource(
    @SerializedName("url")
    val url: String,
    @SerializedName("width")
    val width: Int,
    @SerializedName("height")
    val height: Int
)

/**
 * Model containing information about image resolution
 */
data class ImageResolution(
    @SerializedName("url")
    val url: String?,
    @SerializedName("width")
    val width: Int,
    @SerializedName("height")
    val height: Int
)